physics/discrete_trajectory_body.hpp

```diff
@@ -45,13 +45,14 @@ DiscreteTrajectory<Frame>::back() const {
 template<typename Frame>
 typename DiscreteTrajectory<Frame>::iterator
 DiscreteTrajectory<Frame>::begin() const {
-  return segments_->front().begin();
+  return empty() ? end() : segment_by_left_endpoint_.begin()->second->begin();
 }
 
 template<typename Frame>
 typename DiscreteTrajectory<Frame>::iterator
 DiscreteTrajectory<Frame>::end() const {
-  return segments_->back().end();
+  return iterator(SegmentIterator(segments_.get(), segments_->end()),
```

Comment:
It irks me a bit that we are leaking here details about the representation of the end of the trajectory (`point_ == std::nullopt`).  A possibility would be to have a (private) factory `DiscreteTrajectoryIterator::EndOfLastSegment` with a single parameter (`segment`) which would (1) DCHECK that segment `is_end()` and (2) reset `point_`.  That way the ugly details would stay in `DiscreteTrajectoryIterator`.

Of course, this may have a performance cost, but it may not because there is a lot of inlining happening in these classes.

Reply:
Done. No performance penalty.